README.md
# Twinkle Button

A button plug-in that can flash.

<div align=center>
    <img src='https://github.com/EricerYang/twinkle_button/raw/master/twinkle_button.gif' width='350'>
</div>

## Installation

Add dependency to pubspec.yaml

```bash
dependencies:
  ...
  twinkle_button: ^lastest_version
```

Run in your terminal

```bash
flutter packages get
```

## How to use

```dart
import 'package:twinkle_button/twinkle_button.dart';
```

```dart
TwinkleButton(
  buttonTitle: Text(
    'Subscribe with Free Trial',
    style: TextStyle(
      color: Colors.white,
          fontWeight: FontWeight.w300,
          fontSize: 17.0,
        ),
    ),
    buttonColor: Color(0xff3dce89),
    onclickButtonFunction: () {
      print('hello');
    }
);

```

## Attribute

| Parameter  | Default   | Description |
| :------------ |:---------------:| :-----|
| buttonTitle | null  | Change this value if you what to put pagination in other placeThe text on the button, cannot be omitted. |
| buttonColor | null | Button background color, cannot be omitted. |
| buttonHeight | 50.0 | Height of button. |
| buttonWidth | 200.0 | Width of button. |
| durationTime | 3 | Button blink interval.(seconds) |
| twinkleTime | 300 | Blink time.(milliseconds) |
| highlightColor | Colors.white | Flash effect background color. |
| onclickButtonFunction | null | Events executed by clicking the button, cannot be omitted. |


CHANGELOG.md
## [0.0.3] - 2020-01-08

* Add flash effect color interface.

## [0.0.2] - 2020-01-08

* update readme.

## [0.0.1] - 2020-01-08

* Initial release.

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:twinkle_button_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/twinkle_button_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:twinkle_button/twinkle_button.dart';

void main() {
  const MethodChannel channel = MethodChannel('twinkle_button');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });
}


lib/shimmer.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

///
/// An enum defines all supported directions of shimmer effect
///
/// * [ShimmerDirection.ltr] left to right direction
/// * [ShimmerDirection.rtl] right to left direction
/// * [ShimmerDirection.ttb] top to bottom direction
/// * [ShimmerDirection.btt] bottom to top direction
///
enum ShimmerDirection { ltr, rtl, ttb, btt }

///
/// A widget renders shimmer effect over [child] widget tree.
///
/// [child] defines an area that shimmer effect blends on. You can build [child]
/// from whatever [Widget] you like but there're some notices in order to get
/// exact expected effect and get better rendering performance:
///
/// * Use static [Widget] (which is an instance of [StatelessWidget]).
/// * [Widget] should be a solid color element. Every colors you set on these
/// [Widget]s will be overridden by colors of [gradient].
/// * Shimmer effect only affects to opaque areas of [child], transparent areas
/// still stays transparent.
///
/// [period] controls the speed of shimmer effect. The default value is 1500
/// milliseconds.
///
/// [direction] controls the direction of shimmer effect. The default value
/// is [ShimmerDirection.ltr].
///
/// [gradient] controls colors of shimmer effect.
///
/// [loop] the number of animation loop, set value of `0` to make animation run
/// forever.
///
/// [enabled] controls if shimmer effect is active. When set to false the animation
/// is paused
///
///
/// ## Pro tips:
///
/// * [child] should be made of basic and simple [Widget]s, such as [Container],
/// [Row] and [Column], to avoid side effect.
///
/// * use one [Shimmer] to wrap list of [Widget]s instead of a list of many [Shimmer]s
///
@immutable
class Shimmer extends StatefulWidget {
  final Widget child;
  final Duration period;
  final ShimmerDirection direction;
  final Gradient gradient;
  final int loop;
  final bool enabled;

  const Shimmer({
    Key key,
    @required this.child,
    @required this.gradient,
    this.direction = ShimmerDirection.ltr,
    this.period = const Duration(milliseconds: 1500),
    this.loop = 0,
    this.enabled = true,
  }) : super(key: key);

  ///
  /// A convenient constructor provides an easy and convenient way to create a
  /// [Shimmer] which [gradient] is [LinearGradient] made up of `baseColor` and
  /// `highlightColor`.
  ///
  Shimmer.fromColors({
    Key key,
    @required this.child,
    @required Color baseColor,
    @required Color highlightColor,
    this.period = const Duration(milliseconds: 1500),
    this.direction = ShimmerDirection.ltr,
    this.loop = 0,
    this.enabled = true,
  })  : gradient = LinearGradient(
      begin: Alignment.topLeft,
      end: Alignment.centerRight,
      colors: <Color>[
        baseColor,
        baseColor,
        highlightColor,
        baseColor,
        baseColor
      ],
      stops: const <double>[
        0.0,
        0.35,
        0.5,
        0.65,
        1.0
      ]),
        super(key: key);

  @override
  _ShimmerState createState() => _ShimmerState();

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(DiagnosticsProperty<Gradient>('gradient', gradient,
        defaultValue: null));
    properties.add(EnumProperty<ShimmerDirection>('direction', direction));
    properties.add(
        DiagnosticsProperty<Duration>('period', period, defaultValue: null));
    properties
        .add(DiagnosticsProperty<bool>('enabled', enabled, defaultValue: null));
  }
}

class _ShimmerState extends State<Shimmer> with SingleTickerProviderStateMixin {
  AnimationController _controller;
  int _count;

  @override
  void initState() {
    super.initState();
    _count = 0;
    _controller = AnimationController(vsync: this, duration: widget.period)
      ..addStatusListener((AnimationStatus status) {
        if (status != AnimationStatus.completed) {
          return;
        }
        _count++;
        if (widget.loop <= 0) {
          _controller.repeat();
        } else if (_count < widget.loop) {
          _controller.forward(from: 0.0);
        }
      });
    if (widget.enabled) {
      _controller.forward();
    }
  }

  @override
  void didUpdateWidget(Shimmer oldWidget) {
    if (widget.enabled) {
      _controller.forward();
    } else {
      _controller.stop();
    }
    super.didUpdateWidget(oldWidget);
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _controller,
      child: widget.child,
      builder: (BuildContext context, Widget child) => _Shimmer(
        child: child,
        direction: widget.direction,
        gradient: widget.gradient,
        percent: _controller.value,
        enabled: widget.enabled,
      ),
    );
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }
}

@immutable
class _Shimmer extends SingleChildRenderObjectWidget {
  final double percent;
  final ShimmerDirection direction;
  final Gradient gradient;
  final bool enabled;

  const _Shimmer({
    Widget child,
    this.percent,
    this.direction,
    this.gradient,
    this.enabled,
  }) : super(child: child);

  @override
  _ShimmerFilter createRenderObject(BuildContext context) {
    return _ShimmerFilter(percent, direction, gradient, enabled);
  }

  @override
  void updateRenderObject(BuildContext context, _ShimmerFilter shimmer) {
    shimmer.percent = percent;
    shimmer.enabled = enabled;
  }
}

class _ShimmerFilter extends RenderProxyBox {
  final Paint _clearPaint = Paint();
  final Paint _gradientPaint;
  final Gradient _gradient;
  final ShimmerDirection _direction;
  bool enabled;
  double _percent;
  Rect _rect;

  _ShimmerFilter(this._percent, this._direction, this._gradient, this.enabled)
      : _gradientPaint = Paint()..blendMode = BlendMode.srcIn;

  @override
  bool get alwaysNeedsCompositing => child != null;

  set percent(double newValue) {
    if (newValue == _percent) {
      return;
    }
    _percent = newValue;
    markNeedsPaint();
  }

  @override
  void paint(PaintingContext context, Offset offset) {
    if (child == null) {
      return;
    }
    assert(needsCompositing);

    context.canvas.saveLayer(offset & child.size, _clearPaint);
    context.paintChild(child, offset);

    final double width = child.size.width;
    final double height = child.size.height;
    Rect rect;
    double dx, dy;
    if (_direction == ShimmerDirection.rtl) {
      dx = _offset(width, -width, _percent);
      dy = 0.0;
      rect = Rect.fromLTWH(offset.dx - width, offset.dy, 3 * width, height);
    } else if (_direction == ShimmerDirection.ttb) {
      dx = 0.0;
      dy = _offset(-height, height, _percent);
      rect = Rect.fromLTWH(offset.dx, offset.dy - height, width, 3 * height);
    } else if (_direction == ShimmerDirection.btt) {
      dx = 0.0;
      dy = _offset(height, -height, _percent);
      rect = Rect.fromLTWH(offset.dx, offset.dy - height, width, 3 * height);
    } else {
      dx = _offset(-width, width, _percent);
      dy = 0.0;
      rect = Rect.fromLTWH(offset.dx - width, offset.dy, 3 * width, height);
    }
    if (_rect != rect) {
      _gradientPaint.shader = _gradient.createShader(rect);
      _rect = rect;
    }
    context.canvas.translate(dx, dy);
    context.canvas.drawRect(rect, _gradientPaint);
    context.canvas.restore();
  }

  double _offset(double start, double end, double percent) {
    return start + (end - start) * percent;
  }
}

lib/twinkle_button.dart
import 'dart:async';

///
/// * author: EricerYang
/// * email: y_nan98@163.com
///
/// A package provides a button plug-in that can flash.
///

import 'package:flutter/material.dart';
import 'package:twinkle_button/shimmer.dart';

class TwinkleButton extends StatefulWidget {
  final Text buttonTitle;
  final Color buttonColor;
  final double buttonHeight;
  final double buttonWidth;
  final int durationTime;
  final int twinkleTime;
  final Color highlightColor;
  final GestureTapCallback onclickButtonFunction;

  const TwinkleButton({
    Key key,
    @required this.buttonTitle,
    @required this.buttonColor,
    @required this.onclickButtonFunction,
    this.highlightColor = Colors.white,
    this.buttonHeight = 50.0,
    this.buttonWidth = 280.0,
    this.durationTime = 3,
    this.twinkleTime = 300,
  }) : super(key: key);

  @override
  _TwinkleButtonState createState() => _TwinkleButtonState();
}

class _TwinkleButtonState extends State<TwinkleButton>
    with SingleTickerProviderStateMixin {

  double _scale;
  AnimationController _controller;
  Timer _timer;
  int _twinkleCount;

  void _loopAnimation(int twinkleCount) {
    if(twinkleCount == 1) {
      setState(() {
        _timer = Timer(Duration(seconds: widget.durationTime), _twinkleForward,);
      });
    }
    else {
      _twinkleForward();
    }
  }

  @override
  void initState() {
    super.initState();
    _twinkleCount = 0;
    _controller = AnimationController(
      vsync: this,
      duration: Duration(milliseconds: widget.twinkleTime),
      lowerBound: 0.0,
      upperBound: 0.1,
    )..addListener(() {
      setState(() {});
    })
      ..addStatusListener((listener) {
        if (listener == AnimationStatus.completed) {
          _controller.reverse();
        }
        else if (listener == AnimationStatus.dismissed) {
          if(_twinkleCount == 1) {
            _twinkleCount = 0;
          }
          else {
            _twinkleCount++;
          }
          _loopAnimation(_twinkleCount);
        }
      });
    _loopAnimation(_twinkleCount);
  }

  void _twinkleForward() {
    _controller.forward();
  }

  @override
  void dispose() {
    _controller.dispose();
    _timer?.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    _scale = 1 + _controller.value;

    return GestureDetector(
      onTap: widget.onclickButtonFunction ?? () {},
      child: Transform.scale(
        scale: _scale,
        child: _twinkleButton,
      ),
    );
  }

  Widget get _twinkleButton => Container(
    child: Center(
      child: Stack(
        children: <Widget>[
          Container(
            width: widget.buttonWidth,
            height: widget.buttonHeight,
            child: RaisedButton(
              onPressed: () async {
                return null;
              },
              padding: EdgeInsets.only(
                top: 10.0,
                bottom: 10.0,
                left: 15.0,
                right: 15.0,
              ),
              textColor: Colors.white,
              color: widget.buttonColor,
              shape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(30.0),
              ),
            ),
          ),
          Container(
            width: widget.buttonWidth,
            height: widget.buttonHeight,
            decoration: new BoxDecoration(
              borderRadius: BorderRadius.all(Radius.circular(25.0)),
            ),
            child: Opacity(
              opacity: 0.6,
              child: Shimmer.fromColors(
                baseColor: widget.buttonColor,
                highlightColor: widget.highlightColor,
                child: Container(
                  width: widget.buttonWidth,
                  height: widget.buttonHeight,
                  decoration: new BoxDecoration(
                    color: widget.buttonColor,
                    borderRadius: BorderRadius.all(Radius.circular(25.0)),
                  ),
                ),
              ),
            ),
          ),
          Container(
            width: widget.buttonWidth,
            height: widget.buttonHeight,
            decoration: new BoxDecoration(
              borderRadius: BorderRadius.all(Radius.circular(25.0)),
            ),
            child: Center(
              child: widget.buttonTitle,
            ),
          ),
        ],
      ),
    ),
  );

}


android/src/main/java/com/yn/twinkle_button/TwinkleButtonPlugin.java
package com.yn.twinkle_button;

import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** TwinkleButtonPlugin */
public class TwinkleButtonPlugin implements FlutterPlugin, MethodCallHandler {
  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    final MethodChannel channel = new MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "twinkle_button");
    channel.setMethodCallHandler(new TwinkleButtonPlugin());
  }

  // This static function is optional and equivalent to onAttachedToEngine. It supports the old
  // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
  // plugin registration via this function while apps migrate to use the new Android APIs
  // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
  //
  // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
  // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
  // depending on the user's project. onAttachedToEngine or registerWith must both be defined
  // in the same class.
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "twinkle_button");
    channel.setMethodCallHandler(new TwinkleButtonPlugin());
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:twinkle_button/twinkle_button.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('Twinkle Button'),
        ),
        body: Column(
          children: <Widget>[
            Expanded(
              flex: 1,
              child: Center(
                child: TwinkleButton(
                    buttonTitle: Text(
                      'Download',
                      style: TextStyle(
                        color: Colors.white,
                        fontWeight: FontWeight.w700,
                        fontSize: 17.0,
                      ),
                    ),
                    buttonColor: Color(0xffFF4758),
                    twinkleTime: 100,
                    onclickButtonFunction: () {
                      print('hello');
                    }
                ),
              ),
            ),
            Expanded(
              flex: 1,
              child: Center(
                child: TwinkleButton(
                    buttonTitle: Text(
                      'Subscribe with Free Trial',
                      style: TextStyle(
                        color: Colors.white,
                        fontWeight: FontWeight.w300,
                        fontSize: 17.0,
                      ),
                    ),
                    buttonColor: Color(0xff3dce89),
                    onclickButtonFunction: () {
                      print('hello');
                    }
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
